### ISSUE TITLE
Build broken on macOS Big Sur 11.5.2

### ISSUE THREAD

**Issue by jonas-schulze** [JuliaLang/julia#42346]
Cloning and building Julia on the current master (5bf1ca5c724143091973ef6632f1b3026d8021af)
```bash
git clone ...
make
```
leads to
```
...
    JULIA usr/lib/julia/sys-o.a
/bin/sh: line 1: 11329 Segmentation fault: 11  JULIA_BINDIR=/Users/jonas/Projects/julia2/usr/bin WINEPATH="/Users/jonas/Projects/julia2/usr/bin;$WINEPATH" /Users/jonas/Projects/julia2/usr/bin/julia -O3 -C "native" --output-o /Users/jonas/Projects/julia2/usr/lib/julia/sys-o.a.tmp --startup-file=no --warn-overwrite=yes --sysimage /Users/jonas/Projects/julia2/usr/lib/julia/sys.ji /Users/jonas/Projects/julia2/contrib/generate_precompile.jl 1
*** This error is usually fixed by running `make clean`. If the error persists, try `make cleanall`. ***
make[1]: *** [/Users/jonas/Projects/julia2/usr/lib/julia/sys-o.a] Error 1
make: *** [julia-sysimg-release] Error 2
```
While working on my "old" clone, I encountered the same problem. `make clean` and `make cleanall` didn't fix the issue. I didn't try this on the new clone. I also tried creating a `Make.user` containing `JULIA_PRECOMPILE=0`, but that didn't work either.

I am running macOS 11.5.2 on a 2018 MacBook Pro. From my homebrew Julia installation:
```
julia> versioninfo()
Julia Version 1.6.2
Commit 1b93d53fc4 (2021-07-14 15:36 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.7.0)
  CPU: Intel(R) Core(TM) i5-8259U CPU @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, skylake)
```

**Comment by vtjnash** [comment]
It is likely you have something broken in your environment. Check `env` for something being added.

**Comment by jonas-schulze** [comment]
I don't see anything suspicious. 🙁

```
$ env | cut -d= -f1
TERM_SESSION_ID
SSH_AUTH_SOCK
LC_TERMINAL_VERSION
COLORFGBG
ITERM_PROFILE
SQLITE_EXEMPT_PATH_FROM_VNODE_GUARDS
XPC_FLAGS
LANG
PWD
SHELL
__CFBundleIdentifier
SECURITYSESSIONID
TERM_PROGRAM_VERSION
TERM_PROGRAM
PATH
DISPLAY
LC_TERMINAL
COLORTERM
COMMAND_MODE
TERM
HOME
TMPDIR
USER
XPC_SERVICE_NAME
LOGNAME
LaunchInstanceID
__CF_USER_TEXT_ENCODING
ITERM_SESSION_ID
SHLVL
OLDPWD
EDITOR
TEXINPUTS
JULIA_NUM_THREADS
JULIA_PROJECT
JULIA_PKG_PRECOMPILE_AUTO
PAGER
LESS
LSCOLORS
LC_CTYPE
FZF_DEFAULT_COMMAND
HOMEBREW_GITHUB_API_TOKEN
_
```

**Comment by vtjnash** [comment]
`JULIA_NUM_THREADS` -- the build does not support explicitly launching threads

**Comment by DilumAluthge** [comment]
This seems like an error on our end. If the build doesn't support `JULIA_NUM_THREADS`, then it should ignore it, instead of crashing.

**Comment by DilumAluthge** [comment]
I'm reopening this because I think we need to fix this on our end.

**Comment by vilterp** [comment]
For what it's worth, this happens to me even with `JULIA_NUM_THREADS` unset.

**Comment by NHDaly** [comment]
@vilterp and, it's worth mentioning, it only started happening to you today, on a build that was working up til yesterday.

**Comment by DilumAluthge** [comment]
> @vilterp and, it's worth mentioning, it only started happening to you today, on a build that was working up til yesterday.

It would be great to bisect this.

**Comment by DilumAluthge** [comment]
I am not able to reproduce this locally on macOS Big Sur version 11.6 (20G165).

@jonas-schulze or @vilterp would you be able to `git bisect` this, since you are able reproduce this failure locally?

**Comment by vilterp** [comment]
The thing is, I am now unable to build commits which I was able to build before. It seems like some state on my system is corrupt, which is preventing the build from succeeding.

I tried `make clean`, `make cleanall`, and even deleted my checkout and restarted, but to no avail.

The only thing I can think of which changed from when it was working to now is that I used to only be building the C source, with `make -C src` and `make -j julia-src-release`.

Eventually, I made a change to the Julia source (this is all on https://github.com/JuliaLang/julia/pull/42286), and issued `make`. However, at the time, my C source had a bug in it, so building the sysimage failed. Once I fixed the bug on the C side, I ran into this bug, there the C and sysimage build, but some step afterward segfaults.

Is there some state outside of the julia checkout directory itself which could be corrupted, causing my failed builds?

**Comment by DilumAluthge** [comment]
There shouldn't be any state outside of the Julia source directory.

To be safe, you can try doing e.g. `~/.julia`, but I doubt that will help.

Has anything changed on your system? For example:
1. macOS update?
2. Xcode update?

**Comment by DilumAluthge** [comment]
> I am not able to reproduce this locally on macOS Big Sur version 11.6 (20G165).

Also, if I set e.g. `export JULIA_NUM_THREADS=4` before I do `make`, I still cannot reproduce the build failure locally.

So whatever is going on, I'm not convinced that it's being caused by `JULIA_NUM_THREADS`.

**Comment by DilumAluthge** [comment]
> Has anything changed on your system? For example:
> 
> 1. macOS update?
> 2. Xcode update?

For example, Xcode 13 was released on September 20, 2021. If you install Xcode from the Mac App Store, I think it auto-updates. So maybe that's what happened here?

**Comment by vilterp** [comment]
Hm, maybe it is an XCode update or something.

I pulled out an old laptop, and the same commits build just fine on it.

* old laptop
    * macOS 11.3
    * `clang -v`:

```
clang version 11.0.0
Target: x86_64-apple-darwin20.4.0
Thread model: posix
InstalledDir: /usr/local/opt/llvm/bin
```

* new laptop
    * macOS 11.5.2
    * `clang -v`:

```
Apple clang version 13.0.0 (clang-1300.0.29.3)
Target: x86_64-apple-darwin20.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

Mysterious 🤔

**Comment by DilumAluthge** [comment]
If the Julia build is broken on Xcode 13, that is bad.

**Comment by vilterp** [comment]
Maybe it auto-updated under me. What is the macOS CI running? Will take a look…

**Comment by vilterp** [comment]
Can't tell from this build what version of xcode/clang is running in CI; not sure where the config lives https://build.julialang.org/#/builders/63/builds/3806

**Comment by NHDaly** [comment]
Great sleuthing y'all 👍👍

**Comment by DilumAluthge** [comment]
On my local machine, with otherwise identical build environments, the build passes with Xcode 12 and fails with Xcode 13.

| macOS         | Xcode           | Clang                       | Build              |
| ------------- | --------------- | --------------------------- | -------------------|
| 11.6 (20G165) | 12.5.1 (12E507) | 12.0.5 (clang-1205.0.22.11) | :white_check_mark: |
| 11.6 (20G165) | 13.0 (13A233)   | 13.0.0 (clang-1300.0.29.3)  | :x                 |

```
Sysimage built. Summary:
Total ───────  73.074822 seconds
Base: ───────  29.993381 seconds 41.0448%
Stdlibs: ────  43.078845 seconds 58.9517%
    JULIA usr/lib/julia/sys-o.a
/bin/sh: line 1:  6432 Segmentation fault: 11  JULIA_BINDIR=/Users/dilum/Downloads/julia/usr/bin WINEPATH="/Users/dilum/Downloads/julia/usr/bin;$WINEPATH" /Users/dilum/Downloads/julia/usr/bin/julia -O3 -C "native" --output-o /Users/dilum/Downloads/julia/usr/lib/julia/sys-o.a.tmp --startup-file=no --warn-overwrite=yes --sysimage /Users/dilum/Downloads/julia/usr/lib/julia/sys.ji /Users/dilum/Downloads/julia/contrib/generate_precompile.jl 1
*** This error is usually fixed by running `make clean`. If the error persists, try `make cleanall`. ***
make[1]: *** [/Users/dilum/Downloads/julia/usr/lib/julia/sys-o.a] Error 1
make: *** [julia-sysimg-release] Error 2
```

**Comment by wookay** [comment]
I just found the difference from `git blame` commands between two macs.
ref: https://github.com/JuliaLang/julia/blob/2c9e051/base/version_git.sh#L45

```shell
$ xcodebuild -version
Xcode 13.0
Build version 13A233

$ git --version
git version 2.30.1 (Apple Git-130)

$ git blame -sl VERSION
^5e8d7edfffbac8a69fad0f901089bbaa97d1043 1) 1.8.0-DEV
```

```shell
$ xcodebuild -version
Xcode 12.4
Build version 12D4e

$ git --version
git version 2.24.3 (Apple Git-128)

$ git blame -sl VERSION
1fd46e6251b7e8fd60559ed513144efdbb69cb7c 1) 1.8.0-DEV
```

**Comment by jonas-schulze** [comment]
The last time I worked on Julia (#41834), everything built just fine. Now that commit is "broken", too. I don't know my version of Clang or Xcode back then but now it's 13:

| macOS | Xcode | Clang | Build |
|---|---|---|---|
| Big Sur 11.5.2 (20G95) | 13.0 (13A233) | 13.0.0 (clang-1300.0.29.3) | :boom: |

**Comment by DilumAluthge** [comment]
Yeah, based on the comments from @jonas-schulze and @vilterp, I don't think that anything changed recently in Julia itself. I think the issue is due to Xcode 12 versus Xcode 13.

**Comment by DilumAluthge** [comment]
Yeah, based on the comments from @jonas-schulze and @vilterp, I don't think anything changed in Julia. I think the issue is with Xcode 12 versus Xcode 13.

**Comment by IanButterworth** [comment]
I just hit this but didn't find this issue until after a few hours of head scratching..
I saw no sign that MacOS had updated anything :/
